### PR TITLE
Allow IsPowered to return negative power percentages.

### PR DESCRIPTION
**Pull request by randhid**
I am pretty sure this was not written expecting that PowerPct would always return positive. Just an oversight on taking the absolute value of the power percentage.

Still need to fix tests, but it should be returning negative percentages even when on now.

Please 

- [x] Test on hardware after I rebase - I will add the app image label to this pr.

### CODE REVIEW

File: components/motor/gpio/basic.go [hunk 1]
```diff
@@ -190,35 +195,34 @@ func (m *Motor) setPWM(ctx context.Context, powerPct float64, extra map[string]i
...
-			powerPct = 1.0 - math.Abs(powerPct) // Other pin is always high, so only when PWM is LOW are we driving. Thus, we invert here.
+			pwmPower = 1.0 - pwmPower // Other pin is always high, so only when PWM is LOW are we driving. Thus, we invert here.
 		default:
 			return m.turnOff(ctx, extra)
 		}
 	}
 
-	powerPct = math.Max(math.Abs(powerPct), m.minPowerPct)
-	m.powerPct = powerPct
+	pwmPower = math.Max(pwmPower, m.minPowerPct)
```
Comment (JohnN193) [inline components/motor/gpio/basic.go:221]: `m.powerPct` needs to capture the minPowerPct

File: components/motor/gpio/basic.go [hunk 2]
```diff
@@ -177,8 +177,15 @@ func (m *Motor) setPWM(ctx context.Context, powerPct float64, extra map[string]i
 	var errs error
 	powerPct = math.Min(powerPct, m.maxPowerPct)
 	powerPct = math.Max(powerPct, -1*m.maxPowerPct)
+	powerPct = math.Max(powerPct, m.minPowerPct)
```
Comment (martha-johnston) [inline components/motor/gpio/basic.go:180]: just ran through this quickly with an example of -50% power and I think line 180 forces the powerPct to 0. this is assuming a min power of 0 and a max power of 100, meaning the power range is -100 to +100, assuming I am understanding the min and max powers correctly

File: components/motor/gpio/basic.go [hunk 3]
```diff
@@ -190,35 +197,33 @@ func (m *Motor) setPWM(ctx context.Context, powerPct float64, extra map[string]i
...
-		case powerPct >= 0.001:
+		case pwmPower >= 0.001:
 			pwmPin = m.B
 			if m.dirFlip {
 				pwmPin = m.A
 			}
-			powerPct = 1.0 - math.Abs(powerPct) // Other pin is always high, so only when PWM is LOW are we driving. Thus, we invert here.
-		case powerPct <= -0.001:
+			pwmPower = 1.0 - pwmPower // Other pin is always high, so only when PWM is LOW are we driving. Thus, we invert here.
+		case pwmPower <= -0.001:
```
Comment (martha-johnston) [inline components/motor/gpio/basic.go:212]: isn't there an issue here now that pwmPower is now an absolute value?

Comment (randhid) [reply components/motor/gpio/basic.go:212]: nice catch! I'll ad a minus sign

File: components/motor/gpio/basic.go (hunk 2 above)
Comment (randhid) [reply components/motor/gpio/basic.go:180]: hmmm maybe i should add a greater than check here first.

Comment (randhid) [reply components/motor/gpio/basic.go]: Another thing I could do is move the storage of powerPct and on to the parent SetPower if that makes this function less confusing, but let me check that `setPWM` is not called from anywhere else in `basic.go` later today.

File: components/motor/gpio/basic.go [hunk 4]
```diff
@@ -175,10 +175,20 @@ func (m *Motor) turnOff(ctx context.Context, extra map[string]interface{}) error
 // Anything calling setPWM MUST lock the motor's mutex prior.
 func (m *Motor) setPWM(ctx context.Context, powerPct float64, extra map[string]interface{}) error {
 	var errs error
+
 	powerPct = math.Min(powerPct, m.maxPowerPct)
 	powerPct = math.Max(powerPct, -1*m.maxPowerPct)
+	if math.Abs(powerPct) > m.minPowerPct {
+		powerPct = math.Max(powerPct, m.minPowerPct)
```
Comment (martha-johnston) [inline components/motor/gpio/basic.go:182]: probs can remove this line then

Comment (martha-johnston) [reply components/motor/gpio/basic.go:182]: oh wait I didn't see it was the absolute value. I think this issue will still be there. if the power is -0.5, the if statement is true, but then the max between -0.5 and 0 is still 0

Comment (randhid) [reply components/motor/gpio/basic.go:182]: working on it. most test passing now :)

**Review by martha-johnston — APPROVED** [review]

File: components/motor/gpio/basic.go [hunk 5]
```diff
@@ -186,39 +183,51 @@ func (m *Motor) setPWM(ctx context.Context, powerPct float64, extra map[string]i
 		errs = multierr.Combine(errs, m.EnablePinHigh.Set(ctx, true, extra))
 	}
 
+	m.powerPct = powerPct
+	powerPct = math.Min(powerPct, m.maxPowerPct)
+	powerPct = math.Max(powerPct, -1*m.maxPowerPct)
```
[the remark this answers is not in the document]
Comment (randhid) [reply components/motor/gpio/basic.go:188]: the powerPCt logic in line 205 needs power percent to be negative to flip the AB pins in the SetPower call, and set the PMW on one pin to pwmPct, and the other to the remainder. I get a failing test if i do this because the calculation is wrong and sets the pct to 0 on the wrong pin.

File: components/motor/gpio/basic.go [hunk 6]
```diff
@@ -186,39 +183,51 @@ func (m *Motor) setPWM(ctx context.Context, powerPct float64, extra map[string]i
...
+
```
Comment (JohnN193) [inline components/motor/gpio/basic.go:189]: I think this does what we want. Checks the max & min powers, then updates the motor's m.powerPct

```suggestion
	powerPct = sign(powerPct)*math.Min(math.Abs(powerPct), m.maxPowerPct) // check max power
	powerPct = sign(powerPct)*math.Max(math.Abs(powerPct), m.minPowerPct) // check min power
	m.powerPct = powerPct
```

File: components/motor/gpio/basic.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (martha-johnston) [reply components/motor/gpio/basic.go:188]: ah okay makes sense. do we not care about checking the minimum then?

File: components/motor/gpio/basic.go [hunk 7]
```diff
@@ -186,39 +183,51 @@ func (m *Motor) setPWM(ctx context.Context, powerPct float64, extra map[string]i
...
+	// pin calls on builtin baords will only take poisitve power, so the absolute value is required to be passed
+	// to pwm pins.
+	pwmPower := math.Abs(powerPct)
```
Comment (JohnN193) [inline components/motor/gpio/basic.go:192]: kk so it turns out if we do an abs here, the power isn't negative anymore so the switch statement doesn't work

File: components/motor/gpio/basic.go [hunk 8]
```diff
@@ -186,39 +183,51 @@ func (m *Motor) setPWM(ctx context.Context, powerPct float64, extra map[string]i
...
 			}
-			powerPct = 1.0 - math.Abs(powerPct) // Other pin is always high, so only when PWM is LOW are we driving. Thus, we invert here.
+			pwmPower = 1.0 - pwmPower // Other pin is always high, so only when PWM is LOW are we driving. Thus, we invert here.
 		case powerPct <= -0.001:
 			pwmPin = m.A
 			if m.dirFlip {
 				pwmPin = m.B
 			}
-			powerPct = 1.0 - math.Abs(powerPct) // Other pin is always high, so only when PWM is LOW are we driving. Thus, we invert here.
+			pwmPower = 1.0 - pwmPower // Other pin is always high, so only when PWM is LOW are we driving. Thus, we invert here.
```
Comment (JohnN193) [inline components/motor/gpio/basic.go:221]: this the abs needs to be put back here, otherwise we never enter this case

File: components/motor/gpio/basic.go [hunk 9]
```diff
@@ -186,39 +183,51 @@ func (m *Motor) setPWM(ctx context.Context, powerPct float64, extra map[string]i
...
+	// pin calls on builtin baords will only take poisitve power, so the absolute value is required to be passed
+	// to pwm pins.
+	pwmPower := math.Abs(powerPct)
+
 	var pwmPin board.GPIOPin
 
 	switch m.motorType {
 	case ABPwm, DirectionPwm:
-		if math.Abs(powerPct) <= 0.001 {
+		if pwmPower <= 0.001 {
```
Comment (JohnN193) [inline components/motor/gpio/basic.go:198]: hmm this also needs to be set positive too though

File: components/motor/gpio/basic.go (hunk 8 above)
Comment (randhid) [reply components/motor/gpio/basic.go:221]: the pwmPower is already absolute from powerPct. powerPct is untouched when passed into the function

File: components/motor/gpio/basic.go (hunk 9 above)
Comment (randhid) [reply components/motor/gpio/basic.go:198]: The ABPWM and PWMDir dont' care about the power, but the AB switch statement is switched on powerPCT but only set using pwmPower, which is the absolute of the powerPCt being calcualted

File: components/motor/gpio/basic.go (hunk 7 above)
Comment (martha-johnston) [reply components/motor/gpio/basic.go:192]: I'm fine with that, I want to add this to rpmMonitor in the next refactor so I agree w/ unblocking returning negative.

File: components/motor/gpio/basic.go (hunk 8 above)
Comment (JohnN193) [reply components/motor/gpio/basic.go:221]: i think i just confused myself, the cases all use powerPct still on my local branch

Comment (JohnN193) [reply components/motor/gpio/basic.go:221]: and i got confused by the multiple fields containing power being used

File: components/motor/gpio/basic.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (randhid) [reply components/motor/gpio/basic.go:188]: no since this function has always had a functional minimum of 0.01 now that I'm reading it.

[the remark this answers is not in the document]
Comment (randhid) [reply components/motor/gpio/basic.go:188]: but added that check back, what's going on now is that even if you have a minimum power, the default is to turn off the motor if it's less than 0.001 power in that switch statement. So we have a hard coded minimum.

File: components/motor/gpio/basic.go (hunk 1 above)
Comment (randhid) [reply components/motor/gpio/basic.go:227]: I'm going to move the power calculation up. New PR incoming.

File: components/motor/gpio/basic.go (hunk 9 above)
Comment (JohnN193) [reply components/motor/gpio/basic.go:198]: misread which pwm value was getting used where. my b

File: components/motor/gpio/basic.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (martha-johnston) [reply components/motor/gpio/basic.go:188]: but I assume if the minimum is like 20% all values below that should be forced to 20% since that's what we're doing with the maximum. not just shut off if it's below the minimum?

File: components/motor/gpio/basic.go [hunk 10]
```diff
@@ -214,7 +219,6 @@ func (m *Motor) setPWM(ctx context.Context, powerPct float64, extra map[string]i
 	}
 
 	powerPct = math.Max(math.Abs(powerPct), m.minPowerPct)
```
Comment (JohnN193) [inline components/motor/gpio/basic.go:221]: [nit] I don't think we need this line anymore, since we moved the min logic to the top

**Review by JohnN193 — APPROVED** [review]
LGTM

File: components/motor/gpio/basic.go (hunk 10 above)
Comment (randhid) [reply components/motor/gpio/basic.go:221]: no but I'm not going to touch more in this pr. It will be merged in messy.

Comment (randhid) [reply components/motor/gpio/basic.go:221]: Edit - no I'm msitaken. This is now the only thing making the pwmPinPower positive. otherwise the pin will crash when you set it to negative.

**Review by martha-johnston — APPROVED** [review]
lgtm assuming hardware tests pass
